src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -604,6 +616,29 @@ public static DType AttachDataSourceInfo(DType type, IExternalTabularDataSource
             return returnType;
         }
 
+        public static DType AttachDisplayNameProvider(DType type, DisplayNameProvider displayNames)
+        {
+            type.AssertValid();
+            Contracts.AssertValue(displayNames);
+
+            DType returnType = type.Clone();
+            if (returnType.DisplayNameProvider == null)
+            {
+                returnType.DisplayNameProvider = displayNames;
+                returnType.DisableDisplayNames = false;
+            } 
+            else if (returnType.DisplayNameProvider.Matches(displayNames))
+            {
+                returnType.DisplayNameProvider = displayNames;
+                return returnType;
+            } 
```

Comment:
this is dangerous, given that a) the DNP is mutable, and b) from the caller's perspective, they don't know whether the DNP will end up getting shared or not.

Reply:
Resolved, DNP is not mutateable within the internals of PowerFx.